Fix userallocatedevent imp

maciejbocianski opened the pull request:
### Summary of changes <!-- Required -->

Fix user allocated events implementation:
- store `delay` and `period` inside `UserAllocatedEvent` and set it at every post
- fix user allocated events canceling
- update tests accordingly

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@jamesbeyond @kjbracey-arm 

----------------------------------------------------------------------------------------------------------------

maciejbocianski committed:
user allocated events: fix event cancelling

for user allocated events use id for event state tracking
skip event canceling when already canceled or dispatched

maciejbocianski committed:
UserAllocatedEvent: set delay/period at event posting

Event delay/period can be modified by equeue so it has to be reset at every post.

maciejbocianski committed:
equeue tests update

- test periodic user allocated events
- test user allocated events re-posting
- test unnecessary cancels of user allocated events

maciejbocianski committed:
queue test update

- extend mixed test with periodic events
- improve static event queue test

bulislaw commented on events/source/equeue.c:
```diff
@@ -402,6 +404,10 @@ void equeue_post_user_allocated(equeue_t *q, void (*cb)(void *), void *p)
     unsigned tick = equeue_tick();
     e->cb = cb;
     e->target = tick + e->target;
+    // for user allocated events use event id to track event state
```
That is not very readable, can we use some kind of enum maybe? So we don't introduce magic numbers. Also are you sure that hijacking of this fields is ok (to be fair i'm not 100% sure how it's being used)

maciejbocianski replied:
True, I will use enum.

Hijacking `id` field should be OK - all `user_allocated` `equeue` APIs (and whole equeue implementation) use event address instead event `id` when dealing with user allocated events.

maciejbocianski replied:
@bulislaw I have reviewed `equeue` code once more and found one place where `id` was altered without checking event type. Already fixed see https://github.com/ARMmbed/mbed-os/pull/12319/commits/7215f99d3e61a4564c5c97b5cad0682de517545e

bulislaw commented on events/UserAllocatedEvent.h:
```diff
@@ -260,17 +262,22 @@ class UserAllocatedEvent<F, void(ArgTs...)> {
             return false;
         }
         core_util_atomic_incr_u8(&_post_ref, 1);
+        equeue_event_delay(&_e + 1, _delay);
```
I think that changes the behavior? As before we could change delay and period of already posted events?

maciejbocianski replied:
It wasn't allowed previously. It was guarded by  `MBED_ASSERT(!_post_ref)`.
It isn't a valid use case to change delay/period while dispatching is ongoing.

maciejbocianski committed:
equeue: skip equeue_incid call for user allocated events while dispatching

User allocaded events doesn't utilize id field in a way that normal
event does, so we shouldn't call equeue_incid on it
